### PR TITLE
Docs and some API usabilities

### PR DESCRIPTION
**Pull request by KristofferC**
Fixes https://github.com/JuliaLang/Pkg.jl/issues/422
Fixes https://github.com/JuliaLang/Pkg.jl/issues/332

Comments welcome

### CODE REVIEW

File: src/Types.jl [hunk 1]
```diff
@@ -649,7 +665,7 @@ function parse_package!(ctx, pkg, project_path)
             Pkg.Operations.set_maximum_version_registry!(env, pkg)
         end
     else
-        @warn "package $(pkg.name) at $(project_path) will need to have a [Julia]Project.toml file in the future"
+        # @warn "package $(pkg.name) at $(project_path) will need to have a [Julia]Project.toml file in the future"
```
Comment (KristofferC) [inline src/Types.jl]: This got a bit too spammy with most packages not having project files.

File: src/Pkg.jl [hunk 2]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+
+See also [`PackageSpec`](@ref).
+"""
+const add = API.add
+
+"""
+    Pkg.rm(pkg::Union{String, Vector{String})
+    Pkg.rm(pkg::Union{PackageSpec, Vector{PackageSpec}})
+
+Removes a package from the current project. If the `mode` of `pkg` is
```
Comment (fredrikekre) [inline src/Pkg.jl]: Removes -> Remove

File: src/Pkg.jl [hunk 3]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+See also [`PackageSpec`](@ref), [`PackageMode`](@ref).
+"""
+const rm = API.rm
+
+"""
+    Pkg.update(; level::UpgradeLevel=UPLEVEL_MAJOR, mode::PackageMode = PKGMODE_PROJECT)
+    Pkg.update(pkg::Union{String, Vector{String})
+    Pkg.update(pkg::Union{PackageSpec, Vector{PackageSpec}})
+
+Updates a package `pkg`. If no posistional argument is given, update all packages in the manifest if `mode` is `PKGMODE_MANIFEST` and packages in both manifest and project if `mode` is `PKGMODE_PROJECT`.
```
Comment (fredrikekre) [inline src/Pkg.jl]: Updates -> Update

File: src/Pkg.jl [hunk 4]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+"""
+    Pkg.free(pkg::Union{String, Vector{String})
+    Pkg.free(pkgs::Union{Packagespec, Vector{Packagespec}})
+
+Free a package which removes a `pin` if it exists, or if the package is tracking a path,
+e.g. after [`Pkg.develop`](@ref), go back to tracking registered versions.
+
+## Examples
+
+```
```
Comment (fredrikekre) [inline src/Pkg.jl]: julia

File: src/Pkg.jl [hunk 5]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+"""
+const free = API.free
+
+
+"""
+    Pkg.develop(pkg::Union{String, Vector{String})
+    Pkg.develop(pkgs::Union{Packagespec, Vector{Packagespec}})
+
+Makes a package available for development by tracking it by path.
```
Comment (fredrikekre) [inline src/Pkg.jl]: Makes -> Make

File: src/Pkg.jl [hunk 6]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+If `pkg` is given with only a name or by a URL the packages will be downloaded
+to the location by the environment variable `JULIA_PKG_DEVDIR` with
+`.julia/dev` as the default.
+
+If `pkg` is given as a local path, the package at that path will be tracked.
+
+## Examples
+
+```
```
Comment (fredrikekre) [inline src/Pkg.jl]: julia block, single # and rm empty line

File: src/Pkg.jl [hunk 7]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+
+Update the current manifest with eventual changes to the dependency graph
+from packages that are tracking a path.
+"""
+const resolve = API.resolve
+
+"""
+    Pkg.status(mode::PackageMode=PKGMODE_PROJECT)
+
+Prints out the status of the project/manifest.
```
Comment (fredrikekre) [inline src/Pkg.jl]: Prints -> Print

File: src/Pkg.jl [hunk 8]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+also print for those in the manifest (recursive dependencies).
+"""
+const status = API.status
+
+
+"""
+    Pkg.activate([s::String])
+
+Activate the environment at `s`. The active environment is the environment
+that the package manager modifies when it runs command.
```
Comment (fredrikekre) [inline src/Pkg.jl]: The active environment is the environment that is modified by executing package commands.

File: src/Pkg.jl [hunk 9]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+activate the environment at that path.
+  * If `s` is a non-existing path, activate that path.
+
+
+If no argument is given to `activate`, activate the home project,
+which is the one specified by either `--project` command line when starting julia,
+or `JULIA_PROJECT` environment variable.
+
+# Examples
+
```
Comment (fredrikekre) [inline src/Pkg.jl]: rm empty line

File: src/Pkg.jl [hunk 10]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+```
```
Comment (fredrikekre) [inline src/Pkg.jl]: julia block

File: src/Pkg.jl [hunk 11]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+|:---------------------|:----------------------------------------------|
+| `Package`            | `PackageSpec("Package")`                      |
+| `Package@0.2`        | `PackageSpec(name="Package", version="0.2")`  |
+| `Package=a67d...`    | `PackageSpec(name="Package", uuid="a67d..."`  |
+| `Package#master`     | `PackageSpec(name="Package", rev="master")`   |
+| `local/path#feature` | `PackageSpec(url="local/path"; rev="feature)` |
+| `www.mypkg.com`      | `PackageSpec("url=www.mypkg.com")`              |
+| `--manifest Package` | `PackageSpec(name="Package", mode=PKGSPEC_MANIFEST)`|
+| `--major Package`    | `PackageSpec(name="Package", version=PKGLEVEL_MAJOR`)|
+"""
```
Comment (fredrikekre) [inline src/Pkg.jl]: Very nice with this table!

File: src/Pkg.jl [hunk 12]
```diff
@@ -39,31 +28,310 @@ include("Operations.jl")
...
+julia> Pkg.add(PackageSpec(name="Example", version="0.3")) # Specify version
+
+julia> Pkg.add(PackageSpec(url="https://github.com/JuliaLang/Example.jl", rev="master")) # From url
+
+julia> Pkg.add(PackageSpec(url="/remote/mycompany/juliapackages/OurPackage"))` # From path (has to be a gitrepo)
+```
+
+See also [`PackageSpec`](@ref).
+"""
+const add = API.add
```
[the remark this answers is not in the document]
Comment (KristofferC) [reply src/Pkg.jl]: Yes, `function`.

**Comment by KristofferC** [comment]
Thanks for all the comments, updated!

File: src/Pkg.jl [hunk 13]
```diff
@@ -39,31 +28,303 @@ include("Operations.jl")
...
+"""
+    Pkg.add(pkg::Union{String, Vector{String})
+    Pkg.add(pkg::Union{PackageSpec, Vector{PackageSpec}})
+
+Add a package to the current project. This package will be available using the
+`import` and `using` keywords in the Julia REPL and if the current project is
+a package, also inside that package.
+
+# Examples
+```jl
```
Comment (fredrikekre) [inline src/Pkg.jl]: Does this work or should it be `julia`?

File: src/Pkg.jl [hunk 14]
```diff
@@ -39,31 +28,303 @@ include("Operations.jl")
...
+
+# TODO: decide what to do with this
+const installed = API.installed
+
+"""
+    Pkg.pin(pkg::Union{String, Vector{String})
+    Pkg.pin(pkgs::Union{Packagespec, Vector{Packagespec}})
+
+Pin a package to the current version (or the one given in the `packagespec` or a certain
+git revision. A pinned package is never be updated.
```
Comment (fredrikekre) [inline src/Pkg.jl]: is never be

File: src/Pkg.jl [hunk 15]
```diff
@@ -39,31 +28,303 @@ include("Operations.jl")
...
+
+If `pkg` is given as a local path, the package at that path will be tracked.
+
+# Examples
+```julia
+# By name
+Pkg.develop("Example")
+
+# By url
+Pkg.develop(PackageSpec(url="https://github.com/JuliaLang/Compat.jl", rev="master"))
```
Comment (Nosferican) [inline src/Pkg.jl]: @KristofferC, has this been solved in master?
```
julia> Pkg.develop(PackageSpec(url="https://github.com/JuliaLang/Compat.jl", rev="master"))
ERROR: git revision cannot be given to `develop`
Stacktrace:
 [1] pkgerror(::String) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/Pkg/src/Types.jl:120
 [2] #add_or_develop#13(::Symbol, ::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/Pkg/src/API.jl:39
 [3] #add_or_develop at ./none:0 [inlined]
 [4] #add_or_develop#12 at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/Pkg/src/API.jl:29 [inlined]
 [5] #add_or_develop at ./none:0 [inlined]
 [6] #add_or_develop#10(::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol},NamedTuple{(:mode, :shared),Tuple{Symbol,Bool}}}, ::Function, ::Pkg.Types.PackageSpec) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/Pkg/src/API.jl:27
 [7] #add_or_develop at ./none:0 [inlined]
 [8] #develop#19 at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/Pkg/src/API.jl:70 [inlined]
 [9] develop(::Pkg.Types.PackageSpec) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/Pkg/src/API.jl:70
 [10] top-level scope at none:0
```

Comment (fredrikekre) [reply src/Pkg.jl]: It's not supposed to work, and it is not documented anymore.

Comment (Nosferican) [reply src/Pkg.jl]: Should people just checkout the branch through Git/vcs after `dev` gets the repo?

Comment (fredrikekre) [reply src/Pkg.jl]: Yes.

**Pull request merged**